src/devices/RGBLedMatrix/RgbLedMatrix.cs

```diff
@@ -0,0 +1,641 @@
+        /// </summary>
+        public void Dispose()
+        {
+            if (_controller != null)
+            {
+                StopRendering();
+
+                while (!_safeToDispose)
+                {
+                    Thread.SpinWait(1);
+                }
+
+                _controller.Dispose();
+                _controller = null;
+            }
+        }
+
+        /// <summary>
+        /// Start rendering on the display area.
+        /// </summary>
+        public void StartRendering()
+        {
+            if (Interlocked.CompareExchange(ref _isRendering, 1, 0) != 0)
+            {
+                return; // we already rendering
+            }
+
+            Task.Factory.StartNew(
+                Render,
+                CancellationToken.None,
+                TaskCreationOptions.LongRunning,
+                TaskScheduler.Default);
+        }
+
+        /// <summary>
+        /// Stop rendering on the display area.
+        /// </summary>
+        public void StopRendering() => Interlocked.CompareExchange(ref _isRendering, 0, 1);
+
+        /// <summary>
+        /// Draw a bitmap on the matrix display area
+        /// </summary>
+        /// <param name="x">Upper left x coordinate on the display</param>
+        /// <param name="y">Upper left y coordinate on the display</param>
+        /// <param name="bitmap">System.Drawing.Bitmap object to draw</param>
+        /// <param name="backBuffer">true if want use back buffer, false otherwise</param>
+        public unsafe void DrawBitmap(int x, int y, Bitmap bitmap, bool backBuffer = false)
+        {
+            if (y >= Height || x >= Width || x + bitmap.Width <= 0 || y + bitmap.Height <= 0)
+            {
+                return;
+            }
+
+            byte [] buffer = backBuffer ? _colorsBackBuffer : _colorsBuffer;
+
+            Rectangle fullImageRectangle = new Rectangle(0, 0, bitmap.Width, bitmap.Height);
+            Rectangle partialBitmap = new Rectangle(x, y, bitmap.Width, bitmap.Height);
+            partialBitmap.Intersect(new Rectangle(0, 0, Width, Height));
+
+            BitmapData bitmapData = bitmap.LockBits(fullImageRectangle, ImageLockMode.ReadOnly, PixelFormat.Format24bppRgb);
+
+            int pos = 3 * ((y < 0 ? Math.Abs(y) * bitmap.Width : 0) + (x < 0 ? Math.Abs(x) : 0));
+            int stride = (bitmapData.Stride - 3 * bitmap.Width) + 3 * (bitmap.Width - partialBitmap.Width);
+
+            Span<byte> span = new Span<byte>((void*) bitmapData.Scan0, fullImageRectangle.Width * fullImageRectangle.Height * 3);
+
+            for (int j = 0; j < partialBitmap.Height; j++)
+            {
+                for (int i = 0; i < partialBitmap.Width; i++)
+                {
+                    SetPixel(partialBitmap.X + i, partialBitmap.Y + j, span[pos + 2], span[pos + 1], span[pos], buffer);
+                    pos += 3;
+                }
+
+                pos += stride;
+            }
+
+            bitmap.UnlockBits(bitmapData);
```

Comment:
does it make sense to wrap this in try .. finally?

Reply:
This is a good idea. I'll do that.